formazione/elenchi.py

```diff
@@ -12,7 +12,7 @@ def template(self):
 
     def excel_colonne(self):
         return super(ElencoPartecipantiCorsiBase, self).excel_colonne() + (
-            ("Stato", lambda p: 'Iscritto' if not p.aspirante or self.args[0][0].pk not in p.aspirante.inviti_attivi else 'Invitato'),
+            ("Stato", lambda p: self.args[0][0].pk),
```

Comment:
Non capisco, cosa contiene `self.args[0][0].pk`?

Il metodo `Elenco.excel_colonne` dovrebbe ritornare un elenco di tuple, dove il primo elemento e' il nome della colonna da mostrare nell'excel e il secondo elemento e' una funzione che prende come argomento una persona nell'elenco (`p`) e ritorna il valore di quella cella.

Reply:
Quando si scarica la lista degli iscritti ad un corso self.args[0][0].pk contiene l'id del corso. Nell'elenco di tuple ci sarà quindi tra le altre cose (Stato, id corso). Così facendo nel file viste.py l'ultimo elemento di "persone_colonne" è proprio l'id del corso che verrà poi usato nel blocco try-except (in "partecipazione").

Reply:
> Quando si scarica la lista degli iscritti ad un corso self.args[0][0].pk contiene l'id del corso. Nell'elenco di tuple ci sarà quindi tra le altre cose (Stato, id corso). Così facendo nel file viste.py l'ultimo elemento di "persone_colonne" è proprio l'id del corso che verrà poi usato nel blocco try-except (in "partecipazione").

OK. `Elenco.excel_colonne` deve pero' ritornare una tupla dove il secondo elemento e' una funzione a cui puo' essere dato in pasto l'oggetto persona. 

E' possibile che l'Excel venga generato senza passare dalla vista `us_elenco_download`, come nei casi degli elenchi che vengono generati ed inviati via email.

Inoltre `us_elenco_download` e' un metodo generico che viene utilizzato per tutti gli elenchi, non e' quindi il luogo appropriato per del codice specifico per una sola tipologia di elenchi. Temo che modificare `us_elenco_download` in questo modo possa compromettere il funzionamento di tutti gli altri elenchi di Gaia, dove la colonna "Stato" potrebbe non esistere.

Ti suggerisco di sostituire la funzione anonima lambda con una funzione con un corpo dove puoi andare a inserire la logica che hai attualmente inserito in `us_elenco_download`, ad esempio:

```python
def excel_colonne(self):

    # Trova il corso base
    corso = CorsoBase.objects.get(pk=self.args[0][0].pk)

    # Crea un metodo che puo' essere usato per generare il valore delle celle "Stato".
    def _persona_stato_iscrizione(persona):
        """
        Ottieni una stringa che descrive lo stato di inscrizione di una persona al corso.
       
        :param persona: Un oggetto Persona.
        :return: "Iscritto" se la persona e' iscritta al corso, altrimenti "Non iscritto".
        """"
        invitato = persona.partecipazioni_corsi.filter(corso=corso).exists()
        return "Iscritto" if invitato else "Non iscritto"

    return super(ElencoPartecipantiCorsiBase, self).excel_colonne() + ( 
        ("Stato", _persona_stato_iscrizione),
    )
```

Reply:
Ieri abbiamo avuto la stessa idea e avevamo apportato questa modifica:

    def excel_colonne(self):

        # stato_iscritto ritorna se la persona è iscritta o ivitata al corso.
        def stato_iscritto(value,self):
            try:
                go = value.aspirante
            except :
                go = None
            if(not go or self.args[0][0].pk not in go.inviti_attivi):
                return "Iscritto"
            else:
                return "Invitato"
        return super(ElencoPartecipantiCorsiBase, self).excel_colonne() + (
            ("Stato", lambda p: stato_iscritto(p,self)),
        )

Se vogliamo ottenere lo stato dell'iscritto con la stessa logica usata nel report "formazione_elenchi_inc_iscritti.html" converrebbe usare il codice su riportato. In caso contrario inserisco il codice da te suggerito, fammi sapere.

 `{% block elenco_riga_extra %}
    <td>{% if not persona.aspirante or elenco.args.0.0.pk not in persona.aspirante.inviti_attivi  %}Iscritto{% else %}Invitato{% endif %}</td>
 {% endblock %}`